### PR TITLE
feat(sql): AllowConnectionFailures does not cause Telegraf to exit on initial connection failure - disabled by default

### LINKED ISSUE #11457: [[inputs.sql]] Looping telegraf if one instance is down
### Relevant telegraf.conf

```toml
[[inputs.sql]]
  driver = "mysql"
  dsn = "xxxx:xxxx@tcp(10.0.0.100:3306)/master"
[[inputs.sql.query]]
  query="SELECT SELECT usrGrp AS usr_grp,FROM current_users WHERE endTime = 0"
[[inputs.sql]]
  driver = "mysql"
  dsn = "xxxx:xxxx@tcp(10.0.0.101:3306)/master"
[[inputs.sql.query]]
  query="SELECT SELECT usrGrp AS usr_grp,FROM current_users WHERE endTime = 0"
```

### Logs from Telegraf

```text
! [inputs.sql] Testing connectivity...
! [inputs.sql] Preparing statement "SELECT usrGrp AS usr_grp,FROM current_users WHERE endTime = 0"...
! [inputs.sql] Connecting to "test:test@tcp(10.0.0.100:3306)/master"...
! [inputs.sql] Testing connectivity...
! [telegraf] Error running agent: starting input inputs.sql: connecting to database failed: dial tcp 10.0.101:3306: i/o timeout
! Starting Telegraf 1.23
```

### System info

Telegraf 1.23.0  - Debian 

### Docker

_No response_

### Steps to reproduce

1. Create 2 [inputs.sql] where one instance is down

### Expected behavior

Expected telegraf to start 

### Actual behavior

It just restarts and loops when it can not connect to the instance that is down and telegraf never starts.

If instance goes down while telegraf is running it breaks collecting and it starts to loop again. 

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by conorevans**
TODO: readme, unit tests (?), actually setting `AllowConnectionFailures` (I think? the SQL config is a bit unusual compared to the plugins I've worked with so far)

- [ ] Updated associated README.md.
- [ ] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format

resolves #11457 

If the initial connection fails, we spawn a goroutine to periodically reconnect. I will add README (though I don't think I can add unit tests, maybe that's just a lack of knowledge though). This is an initial implementation so would like to get some feedback before continuing... it seems sound to me but I'm not ultra experienced in Go - to my knowledge once we spawn the goroutine the recursive calls therein should be handled within that same goroutine (i.e. not spawn additional goroutines that need cleaning, and won't block the main program)

### CODE REVIEW

**Comment by conorevans** [comment]
@powersj if you don't mind taking a look, thanks 🙏

File: plugins/inputs/sql/sql.go [hunk 1]
```diff
@@ -357,10 +368,54 @@ func (s *SQL) Start(_ telegraf.Accumulator) error {
 	// Connect to the database server
 	s.Log.Debugf("Connecting to %q...", s.Dsn)
 	s.db, err = dbsql.Open(s.driverName, s.Dsn)
+	if err != nil {
+		if !s.AllowConnectionFailures {
+			return err
+		}
+		// if connection failures allowed, periodically attempt to reconnect to the DB
+		go s.connect()
```
Comment (conorevans) [inline plugins/inputs/sql/sql.go:376]: Should also return here to not start below (start is called from within connect)

**Comment by Hipska** [comment]
Could you already fix the linter errors?

**Comment by powersj** [comment]
@conorevans sorry I never responded on this one. Not sure how I missed this. I started looking at this today, rebased on master to at least get things passing, but it isn't in a working state. I need to sync with Sven on how we want to do this. You certainly have the right idea and I think what we need to work out is how we want to do this in general for these service inputs. 

Part of me thinks each service input may be unique in how it retries (e.g. kafka recently had retries added in a very specific way). On the other hand this could be as simple as a boolean that says to ignore connection errors and retries at each gather cycle.

Let me sync up and get back to looking at this again.

File: plugins/inputs/sql/sql.go [hunk 2]
```diff
@@ -204,16 +204,26 @@ func (q *Query) parse(acc telegraf.Accumulator, rows *dbsql.Rows, t time.Time) (
 	return rowCount, nil
 }
 
+// If users allow connection failures, we will attempt to reconnect
+// every at each reconnectInterval.
+// We'll apply exponential backoff up until a sane max backoff
+// of 60s, the default Agent.Interval at which each Gather is run.
+var (
+	reconnectInterval    = 500 * time.Millisecond
+	maxReconnectInterval = 60 * time.Second
```
Comment (Hipska) [inline plugins/inputs/sql/sql.go:213]: `maxReconnectInterval` can be a constant.

**Review by Hipska — CHANGES_REQUESTED** [review]

**Pull request closed without merging**